common/src/main/java/org/conscrypt/CryptoUpcalls.java

```diff
@@ -130,7 +122,17 @@ private CryptoUpcalls() {}
         }
     }
 
+    static byte[] rsaSignDigestWithPrivateKey(PrivateKey javaKey, int openSSLPadding,
+            byte[] message) {
+        return rsaOpWithPrivateKey(javaKey, openSSLPadding, Cipher.ENCRYPT_MODE, message);
```

Comment:
Based on our tests, it works when BoringSSL talks to itself.  Possibly BoringSSL is more lenient than is necessary when removing the padding?  I'll switch it back to using Signature for PKCS#1 padding, since it sounds like that's more correct.

Reply:
I wonder if `Cipher.ENCRYPT_MODE` is hooked up funny then. We definitely will not accept the encryption padding in signature verification.

Reply:
It looks like our Cipher and our Signature implementation are implemented basically the same, yep: call `RSA_private_encrypt(..., key, RSA_PKCS1_PADDING)`.  What's the right call to make?

Reply:
Ah, okay. Yeah, that's the signature padding, despite being called "encrypt". The naming is terrible. :-)

For an RSA encryption scheme, you encrypt with the public key and decrypt with the private key, hence `RSA_public_encrypt` and `RSA_private_decrypt`. With a signature scheme, you "encrypt" with the private key and "decrypt" with the public key. Hence `RSA_private_encrypt` and `RSA_public_decrypt`. (We now provide slightly more sensible equivalent APIs with more `size_t`-friendly calling conventions.)

So I guess then the question is whether Java believes `Cipher.ENCRYPT_MODE` + RSA private key + PKCS#1 is supposed to properly encrypt with the public half of the RSA private key, or if it's supposed to do the signature thing because you've got a private key and you are "encrypting". If the latter, this code is correct and I misunderstood the Java API. If the former, both this code and the implementation for RSA keys probably wants to change? (Though that may not be a backwards-compatible change at this point...)

(PS: It's the holidays in the states so I probably won't look at this until Friday. :-) )

Reply:
Ah, yeah, this is what the Java API does.  `Cipher` + `ENCRYPT_MODE` + RSA private key = sign (and `Cipher` + `DECRYPT_MODE` + RSA public key = "unsign", like `RSA_public_decrypt` does).  Sounds like this should be okay, then.